src/languages/delphi.js

```diff
@@ -164,20 +164,28 @@ export default function(hljs) {
     relevance: 0,
     // Source: https://www.freepascal.org/docs-html/ref/refse6.html
     variants: [
+      {
+        // Regular numbers, e.g., 123, 123.456.
+        begin: /\b\d[\d_]*(\.\d[\d_]*)?/ },
       {
         // Hexadecimal notation, e.g., $7F.
-        begin: '\\$[0-9A-Fa-f]+' },
+        begin: /\$[\dA-Fa-f_]*/ },
       {
         // Octal notation, e.g., &42.
-        begin: '&[0-7]+' },
+        begin: /&[0-7][0-7_]*/ },
       {
         // Binary notation, e.g., %1010.
-        begin: '%[01]+' }
+        begin: /%[01_]*/ }
     ]
   };
   const CHAR_STRING = {
     className: 'string',
-    begin: /(#\d+)+/
+    variants: [
+      { begin: /#\d[\d_]*/ },
+      { begin: /#\$[\dA-Fa-f][\dA-Fa-f_]*/ },
+      { begin: /#&[0-7][0-7_]*/ },
+      { begin: /#%[01][01_]*/ }
```

Comment:
Please use `match` since there is no `end` regex here...

Reply:
Sure thing! I'll change that.